### PR TITLE
[core][distributed] fix ray worker rank assignment

### PR DESCRIPTION
**Pull request by youkaichao**
looks like ray does not guarantee the ranks align with machine boundary. when this happens, tensor parallel will happen across machines, which will bring terrible performance.

### CODE REVIEW

File: vllm/executor/ray_gpu_executor.py [hunk 1]
```diff
@@ -134,11 +134,32 @@ def _init_workers_ray(self, placement_group: "PlacementGroup",
...
-        node_workers = defaultdict(list)
-        node_gpus = defaultdict(list)
-
-        for i, (node_id, gpu_ids) in enumerate(worker_node_and_gpu_ids):
-            node_workers[node_id].append(i)
+        # the order in `worker_node_and_gpu_ids` does not necessarily match
+        # the machine boundaries. We need to make sure that workers in the
+        # same node are assigned consecutive ranks.
+        # examples:
+        # [('852a09a13c7503ef126d7c828454c741494b1be33a8627a5206604d9', [0]), ('dfaad7adfdae57a694cc74490db45bd112c9f31243523e43ddc2e7f0', [0]), ('dfaad7adfdae57a694cc74490db45bd112c9f31243523e43ddc2e7f0', [1]), ('dfaad7adfdae57a694cc74490db45bd112c9f31243523e43ddc2e7f0', [2]), ('dfaad7adfdae57a694cc74490db45bd112c9f31243523e43ddc2e7f0', [3]), ('852a09a13c7503ef126d7c828454c741494b1be33a8627a5206604d9', [1]), ('852a09a13c7503ef126d7c828454c741494b1be33a8627a5206604d9', [2]), ('852a09a13c7503ef126d7c828454c741494b1be33a8627a5206604d9', [3])] # noqa
```
Comment (andoorve) [inline vllm/executor/ray_gpu_executor.py:141]: Are we guaranteed that the GPU IDs for a given node will be in strictly increasing order?

Comment (youkaichao) [reply vllm/executor/ray_gpu_executor.py:141]: I don't think so. this is not needed, either. the `CUDA_VISIBLE_DEVICES` is `",".join(map(str, node_gpus[node_id]))`, and local rank is `node_workers[node_id].index(rank)`, as long as these two are consistent, it should be fine.

`node_gpus[node_id]`: list of gpu ids for each worker in this node
`node_workers[node_id]`: list of worker ranks for each worker in this node

**Review by andoorve — APPROVED** [review]

File: vllm/executor/ray_gpu_executor.py [hunk 2]
```diff
@@ -134,11 +134,32 @@ def _init_workers_ray(self, placement_group: "PlacementGroup",
...
+
+        # initialize worker ranks with -1 (unassigned)
+        worker_ranks = [-1 for x in worker_node_and_gpu_ids]
```
Comment (andoorve) [inline vllm/executor/ray_gpu_executor.py:144]: You can do this simply by using python sort/sorted, but this can be very unreadable.

Comment (youkaichao) [reply vllm/executor/ray_gpu_executor.py:144]: I wish ray can sort by node & GPU id for me :( but it cannot

**Pull request merged**